### DISCUSSION TITLE
Haidi BMS to Auxsol inverter Integration

### DISCUSSION THREAD

**Discussion by RestartDK** [ai-republic/bms-to-inverter#33] [Q&A] [answered]
Hello, would it be possible to connect a **Haidi BMS** system to **Auxsol inverter** using CAN? I have had issues recently trying to control parameters from the Haidi system with Auxsol.

**Comment by ai-republic** [comment]
Hi Daniel,
could you give me some information on what protocols the BMS and inverter support? Great would be documentation about the CAN communication protocols but most of the chinese BMS and inverters usually support the PylonTech CAN protocol which is supported (PYLON_CAN & PYLON_HV_CAN). [HV means High Voltage usually >=48V].
Also please specify exact models so that I might research a little ;)

Oh, and if you already have hardware to read out CAN messages, e.g. PI with a CAN hat, you could maybe just dump the CAN messages and I might be able to see which protocol its using.

**Reply by RestartDK** [reply]
Hi,

Could I send you the details for the BMS and inverter documents via email? In short the specific models are Auxsol ASG-(3.6-6)SL-ZH and Haidi 20KW/50KWH AiO Integrated High Voltage Storage System. The communication network between the inverter and the BMS adopts the CAN2.0B communication protocol. I am getting a CAN hat for the raspberry pi right now.

**Reply by ai-republic** [reply]
Just attach the files here, then its also easier for me as a reference. Thanks :)

**Reply by RestartDK** [reply]
[20KW50KWH Haidi AiO Operation Manual.pdf](https://github.com/user-attachments/files/16009896/20KW50KWH.Haidi.AiO.Operation.Manual.pdf)
[QSX J03.08.135-2023 Hybrid inverter and BMS communication protocol (1).pdf](https://github.com/user-attachments/files/16009897/QSX.J03.08.135-2023.Hybrid.inverter.and.BMS.communication.protocol.1.pdf)
[Product+information(HYBRID+INVERTER+3.6-12kW） (3).pdf](https://github.com/user-attachments/files/16009898/Product%2Binformation.HYBRID%2BINVERTER%2B3.6-12kW.3.pdf)

Here I have attached the manuals for each one. One for the models and another explaining the communication protocol

**Reply by RestartDK** [reply]
Also When I use the program do I need to be connected to my inverter or can I just connect to the BMS?

**Reply by ai-republic** [reply]
So the good news is that the AuxSol inverter is compatible with the Pylon HV (_PYLON_HV_CAN_) protocol binding. 

I'm guessing you want to use just the battery system of the HAIDI system. From the looks of it, it looks very similar to the Pylon Powercube battery system. But I don't know if it is also compatible with the Pylon HV CAN protocol. Most chinese products are, but I would either need a CAN communication specification document or a CAN dump to check.

You can just connect the BMSes. If you don't want to connect an inverter just select the _NONE_ inverter binding.

**Reply by RestartDK** [reply]
Here is a CAN dump file from the BMS/ I am having trouble finding a BMS CAN documentation for the HAIDI system:

[dump.txt](https://github.com/user-attachments/files/16051004/dump.txt)

**Comment by ai-republic** [comment]
ok, great! Thanks for the CAN dump. Couldl you do that again and also share the values shown for the BMS like SOC, Voltage, Current, Temperature, etc.? Is that possible? A photo, screenshot or just notes will do, But it must corellate to the CAN dump, so both should be done at the same time

**Reply by RestartDK** [reply]
[program-dump.zip](https://github.com/user-attachments/files/16097391/program-dump.zip)

Attached I have a video of me using the program with a can dump and the actual program running

**Reply by ai-republic** [reply]
Ok, I will check that tomorrow :)

**Reply by ai-republic** [reply]
Ah, you made a video of the CAN dump and the application. 
What I meant, that I need the CAN dump and the states of your battery system, like SOC, SOH, Cell voltages, etc. that you might read out with the manufacturer's provided application. That way I could see what the CAN commands 500, 501, 600, 601 are for and determine how their data is encoded.

**Reply by RestartDK** [reply]
[candump.txt](https://github.com/user-attachments/files/16233338/candump.txt)
![example](https://github.com/user-attachments/assets/355f8c1e-80de-4f96-8db3-c2922de925bf)

Here I have made a candump with the picture of the BMS system

**Reply by RestartDK** [reply]
So I got the CAN documentation from HAIDI now and it has the same commands as the PYLON_HV one. However, when I would test it I think I had the problem of not setting the correct baudrate. The HAIDI system has a baudrate of 250 kb/s and PYLON_HV is set as 500 kb/s. I am trying to make my own module with a different baudrate but I am unfamiliar with using a maven project... Could you please aid me in finding out how to do this?

**Comment by ai-republic** [comment]
Ok that sounds great. Hmm, about the baudrate it would be in the `PylonHVInverterCANDescriptor` class. But in general maybe I should make this configurable in the _Configurator_ application. I can squeeze that in, in the next days. I'll let you know :)

**Comment by ai-republic** [comment]
Actually, I just implemented it so you can set baudrates on BMSes and inverter :)
Please load your configuration in the _Configurator_ application and edit your BMS and inverter then click _Update Configuration_.

**Reply by RestartDK** [reply]
Thank you so much for making the baudrate changeable! Now the application works however I am still having problems with displaying the correct information... Below I have provided the CAN BUS protocol of the system I am working on. Whenever I run the application the SOC value is empty. 
[CAN-Bus-Protocol-high-voltag-V1.17- 高压通讯协议1.pdf](https://github.com/user-attachments/files/16411184/CAN-Bus-Protocol-high-voltag-V1.17-.1.pdf)

**Comment by ai-republic** [comment, marked as answer]
@RestartDK  Great!
Thanks for the communication specification. The specification is idententical to the PylonTech HV specification so it should work.
Could you attach the log files here (running with debug log-level) and maybe a screenshot of the values of the battery application? Maybe we need to adjust the ID for the BMS.
